transport-native-io_uring/src/main/c/netty_io_uring_native.c

```diff
@@ -0,0 +1,525 @@
+  sq->sqes = mmap(0, size, PROT_READ | PROT_WRITE, MAP_SHARED | MAP_POPULATE,
+                  fd, IORING_OFF_SQES);
+  if (sq->sqes == MAP_FAILED) {
+    ret = -errno;
+  err:
+    io_uring_unmap_rings(sq, cq);
+    return ret;
+  }
+
+  cq->khead = cq->ring_ptr + p->cq_off.head;
+  cq->ktail = cq->ring_ptr + p->cq_off.tail;
+  cq->kring_mask = cq->ring_ptr + p->cq_off.ring_mask;
+  cq->kring_entries = cq->ring_ptr + p->cq_off.ring_entries;
+  cq->koverflow = cq->ring_ptr + p->cq_off.overflow;
+  cq->cqes = cq->ring_ptr + p->cq_off.cqes;
+  return 0;
+}
+
+void setup_io_uring(int ring_fd, struct io_uring *io_uring_ring,
+                    struct io_uring_params *p) {
+  int ret;
+
+  ret = io_uring_mmap(ring_fd, p, &io_uring_ring->sq, &io_uring_ring->cq);
+  if (!ret) {
+    io_uring_ring->flags = p->flags;
+    io_uring_ring->ring_fd = ring_fd;
+  } else {
+    perror("setup_io_uring error \n");
+  }
+}
+
+void io_uring_prep_write(struct io_uring_sqe *sqe, int fd, const void *buf,
+                         unsigned nbytes, off_t offset) {
+  io_uring_prep_rw(IORING_OP_WRITE, sqe, fd, buf, nbytes, offset);
+}
+
+void io_uring_prep_read(struct io_uring_sqe *sqe, int fd, void *buf,
+                        unsigned nbytes, off_t offset) {
+  io_uring_prep_rw(IORING_OP_READ, sqe, fd, buf, nbytes, offset);
+}
+
+void io_uring_sqe_set_data(struct io_uring_sqe *sqe, unsigned long data) {
+  sqe->user_data = (unsigned long)data;
+}
+
+void queue_read(int file_fd, struct io_uring *ring, void *buffer, jint event_id,
+                jint pos, jint limit) {
+  struct io_uring_sqe *sqe = NULL;
+
+  sqe = io_uring_get_sqe(ring);
+  if (!sqe) {
+    fprintf(stderr, "Could not get SQE.\n");
+    return;
+  }
+  io_uring_prep_read(sqe, file_fd, buffer + pos, (size_t)(limit - pos), 0);
+  io_uring_sqe_set_data(sqe, (int)event_id);
+}
+
+void queue_write(int file_fd, struct io_uring *ring, void *buffer,
+                 jint event_id, jint pos, jint limit) {
+  struct io_uring_sqe *sqe;
+
+  sqe = io_uring_get_sqe(ring);
+  if (!sqe) {
+    fprintf(stderr, "Could not get SQE.\n");
+    return;
+  }
+  io_uring_prep_write(sqe, file_fd, buffer + pos, (size_t)(limit - pos), 0);
+  io_uring_sqe_set_data(sqe, (unsigned long)event_id);
+}
+
+int __io_uring_peek_cqe(struct io_uring *ring, struct io_uring_cqe **cqe_ptr) {
+  struct io_uring_cqe *cqe;
+  unsigned head;
+  int err = 0;
+
+  do {
+    io_uring_for_each_cqe(ring, head, cqe) break;
+    break;
+  } while (1);
+
+  *cqe_ptr = cqe;
+  return err;
+}
+
+long io_uring_wait_cqe(struct io_uring *ring, unsigned wait_nr) {
+  struct io_uring_cqe *cqe = NULL;
+  int ret = 0, err;
+  unsigned flags = 0;
+
+  err = __io_uring_peek_cqe(ring, &cqe);
+  if (err) {
+    printf("error peek \n");
+    return -errno;
+  }
+
+  if (cqe) {
+    return (long)cqe;
+  }
+
+  flags = IORING_ENTER_GETEVENTS;
+  ret = sys_io_uring_enter(ring->ring_fd, 0, wait_nr, flags, NULL);
```

Comment:
I think this `io_uring_wait_cqe` and `io_uring_submit` should be one operation. The `io_uring_submit` function sets `min_complete` to zero, and this one sets `to_submit` to zero, so likely the control loop would have to make two system calls to make meaningful progress. If you could do both in one call, like how `io_uring_enter` itself is designed, then we could amortise the call overhead even more. This way, I imagine the event loop as basically a control loop for managing the queues, if that makes sense.